test/src/SpendPermissions/approveWithRevoke.t.sol

```diff
@@ -0,0 +1,403 @@
+            mockSpendPermissionManager.getLastUpdatedPeriod(existingSpendPermission);
+        vm.startPrank(address(account));
+        vm.expectRevert(
+            abi.encodeWithSelector(
+                SpendPermissionManager.InvalidLastUpdatedPeriod.selector,
+                invalidLastUpdatedPeriod,
+                lastValidUpdatedPeriod
+            )
+        );
+        mockSpendPermissionManager.approveWithRevoke(
+            newSpendPermission, existingSpendPermission, lastValidUpdatedPeriod
+        );
+        vm.stopPrank();
+    }
+
+    function test_approveWithRevoke_revert_invalidLastUpdatedPeriod_sameSpendSeveralElapsedPeriods(
+        address spender,
+        uint48 start,
+        uint48 end,
+        uint48 period,
+        uint160 allowance,
+        uint256 salt,
+        bytes memory extraData
+    ) public {
+        vm.assume(spender != address(0));
+        vm.assume(spender != address(account));
+        vm.assume(start > 0);
+        vm.assume(uint256(start) + uint256(period) * 4 < end); // make sure we have room for at least two periods
+        vm.assume(period > 0);
+        vm.assume(period <= (end - start) / 4);
+        vm.assume(allowance > 0);
+        vm.deal(address(account), 1 ether);
+
+        SpendPermissionManager.SpendPermission memory newSpendPermission = SpendPermissionManager.SpendPermission({
+            account: address(account),
+            spender: spender,
+            token: NATIVE_TOKEN,
+            start: start,
+            end: end,
+            period: period,
+            allowance: allowance,
+            salt: salt,
+            extraData: extraData
+        });
+        // spend a little
+        vm.startPrank(existingSpendPermission.spender);
+        mockSpendPermissionManager.spend(existingSpendPermission, 1 wei);
+        lastValidUpdatedPeriod = mockSpendPermissionManager.getLastUpdatedPeriod(existingSpendPermission);
+
+        // jump forward several periods and spend same amount again
+        vm.warp(existingSpendPermission.start + existingSpendPermission.period * 4);
+        mockSpendPermissionManager.spend(existingSpendPermission, 1 wei);
+
+        // this will now be the last updated period onchain
+        SpendPermissionManager.PeriodSpend memory invalidLastUpdatedPeriod =
+            mockSpendPermissionManager.getLastUpdatedPeriod(existingSpendPermission);
+        vm.startPrank(address(account));
+        vm.expectRevert(
+            abi.encodeWithSelector(
+                SpendPermissionManager.InvalidLastUpdatedPeriod.selector,
+                invalidLastUpdatedPeriod,
+                lastValidUpdatedPeriod
+            )
+        );
+        mockSpendPermissionManager.approveWithRevoke(
+            newSpendPermission, existingSpendPermission, lastValidUpdatedPeriod
+        );
+        vm.stopPrank();
+    }
+
+    function test_approveWithRevoke_revert_mismatchedAccounts(
```

Comment:
nit: usually like to put all the revert cases up top in the order they appear in the code so this should come immediately after the invalid sender check